# PROJECT:

In [4]:
import pandas as pd
import biogeme.database as db
import biogeme.biogeme as bio
from biogeme.expressions import Beta, Variable
from biogeme import models
from biogeme import results as res

In [5]:
df = pd.read_table('lpmc11.dat')
database = db.Database('lpmc', df)
df

trip_id  household_id  person_n  trip_n  travel_mode  purpose  fueltype  \
0           1             0         0       1            4        3         1   
1          13             1         1       1            4        3         1   
2          19             4         0       1            3        3         6   
3          20             5         1       0            4        3         1   
4          39             9         2       0            4        3         1   
...       ...           ...       ...     ...          ...      ...       ...   
4995    81035         17603         1       1            3        1         1   
4996    81040         17604         2       0            3        1         1   
4997    81045         17605         0       1            4        3         1   
4998    81066         17608         0       3            3        3         6   
4999    81073         17610         1       0            4        3         2   

      faretype  bus_scale  survey_year  ...  dur_pt_access  dur_pt_rail  \
0            1        1.0            1  ...       0.109444     0.000000   
1            5        0.0            1  ...       0.241389     0.000000   
2            5        0.0            1  ...       0.222500     0.000000   
3            5        0.0            1  ...       0.381667     0.000000   
4            5        0.0            1  ...       0.146944     0.000000   
...        ...        ...          ...  ...            ...          ...   
4995         1        1.0            3  ...       0.278056     0.216667   
4996         2        0.0            3  ...       0.264444     0.000000   
4997         5        0.0            3  ...       0.128889     0.000000   
4998         1        1.0            3  ...       0.092222     0.000000   
4999         1        1.0            3  ...       0.206111     0.000000   

      dur_pt_bus  dur_pt_int  pt_interchanges  dur_driving  cost_transit  \
0       0.055556    0.000000                0     0.059444           1.5   
1       0.122222    0.000000                0     0.132222           0.0   
2       0.312222    0.000000                0     0.221667           0.0   
3       0.062222    0.000000                0     0.117222           0.0   
4       0.225000    0.000000                0     0.200833           0.0   
...          ...         ...              ...          ...           ...   
4995    0.000000    0.000000                0     0.392222           2.4   
4996    0.353333    0.176667                1     0.288889           0.0   
4997    0.045833    0.000000                0     0.067778           0.0   
4998    0.389444    0.000000                0     0.193333           1.5   
4999    0.400000    0.000000                0     0.259167           1.5   

      cost_driving_fuel  cost_driving_ccharge  driving_traffic_percent  
0                  0.15                   0.0                 0.112150  
1                  0.50                   0.0                 0.065126  
2                  0.56                   0.0                 0.086466  
3                  0.41                   0.0                 0.097156  
4                  0.48                   0.0                 0.378976  
...                 ...                   ...                      ...  
4995               0.98                   0.0                 0.547450  
4996               0.85                   0.0                 0.175000  
4997               0.17                   0.0                 0.024590  
4998               0.61                   0.0                 0.485632  
4999               0.88                   0.0                 0.233655  

[5000 rows x 32 columns]

# Model 0

**Variables:**

In [6]:
dur_walking = Variable('dur_walking') #time for walking
dur_cycling = Variable('dur_cycling') #time for cycling

dur_pt_access = Variable('dur_pt_access')
dur_pt_rail = Variable('dur_pt_rail')
dur_pt_bus = Variable('dur_pt_bus')
dur_pt_int = Variable('dur_pt_int')

dur_pt = dur_pt_access + dur_pt_rail + dur_pt_bus + dur_pt_int #total time for public transport
dur_driving = Variable('dur_driving')

cost_transit = Variable('cost_transit') #cost for public transport
cost_driving_fuel = Variable('cost_driving_fuel')
cost_driving_ccharge = Variable('cost_driving_ccharge')

cost_driving = cost_driving_fuel + cost_driving_ccharge #total cost for driving

travel_mode = Variable('travel_mode') #Choice

**Parameters:**

In [7]:
constant_walking = Beta('constant_walking', 0, None, None, 0)
constant_cycling = Beta('constant_cycling', 0, None, None, 0)
constant_pt = Beta('constant_pt', 0, None, None, 0)
beta_cost = Beta('beta_cost', 0, None, None, 0)
beta_time = Beta('beta_time', 0, None, None, 0)

In [8]:
v_walk_Model0 = (
    constant_walking
    + beta_time * dur_walking
)

v_cycling_Model0 = (
    constant_cycling
    + beta_time * dur_cycling
)

v_PT_Model0 = (
    constant_pt
    + beta_cost * cost_transit
    + beta_time * dur_pt
)

v_driving_Model0 = (
    beta_cost * cost_driving
    + beta_time * dur_driving
)

In [9]:
V_Model0 = {1: v_walk_Model0, 2: v_cycling_Model0, 3: v_PT_Model0, 4: v_driving_Model0}
logprob_Model0 = models.loglogit(V_Model0, None, travel_mode)
biogeme_Model0 = bio.BIOGEME(database, logprob_Model0) #create the biogeme estimator
biogeme_Model0.modelName = 'logit_Model0'
res_Model0 = biogeme_Model0.estimate()

In [10]:
print(res_Model0.printGeneralStatistics())

Number of estimated parameters:	5
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4676.875
Final log likelihood:	-4676.875
Likelihood ratio test for the init. model:	-0
Rho-square for the init. model:	0
Rho-square-bar for the init. model:	-0.00107
Akaike Information Criterion:	9363.751
Bayesian Information Criterion:	9396.337
Final gradient norm:	2.6771E-04
Nbr of threads:	8



In [11]:
res_Model0.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
beta_cost        -0.156585      0.012667   -12.361233           0.0
beta_time        -5.226955      0.191905   -27.237207           0.0
constant_cycling -2.485329      0.086252   -28.814839           0.0
constant_pt       0.726783      0.047157    15.411971           0.0
constant_walking  1.187012      0.078799    15.063738           0.0

The sign of the 2 generic coefficent leaves us satisfied. We know it's reasonable that utility is decreased when cost and/or travel time increase. It also seems that time impacts more the utility, since in asbolute value beta_time is higher than beta_cost. Concernig the ASCs, we normalized with respect to ASC_car so this is our baseline. We notice that ASC_walking is the highest so it seems that people, assuiming the same atrributes values, would choose this alternative. Anyway we see that, since the value of beta_time is high in absoulute value, and walking usually requires more time than the others, it will be penalized, as we expected. Cycling seems really disadvantaged, since its ASC is already low, and it has the same beta_time (due to generic coefficent assumption). Significance of value is confirmed by the t-test values that are all above the 1.96 threshold

# Model 1

We decide to split beta_time into 4 alternative specific parameters. Our choice is due to the fact that this coefficent is present in every utility function, and we think it can undergo a wider variation, since time perception is strictly related to the chosen travel mode.

In [12]:
constant_walking = Beta('constant_walking', 0, None, None, 0)
constant_cycling = Beta('constant_cycling', 0, None, None, 0)
constant_pt = Beta('constant_pt', 0, None, None, 0)
beta_cost = Beta('beta_cost', 0, None, None, 0)
beta_time_walking = Beta('beta_time_walking', 0, None, None, 0)
beta_time_cycling = Beta('beta_time_cycling', 0, None, None, 0)
beta_time_pt = Beta('beta_time_pt', 0, None, None, 0)
beta_time_driving = Beta('beta_time_driving', 0, None, None, 0)

In [13]:
v_walk_Model1 = (
    constant_walking
    + beta_time_walking * dur_walking
)

v_cycling_Model1 = (
    constant_cycling
    + beta_time_cycling * dur_cycling
)

v_PT_Model1 = (
    constant_pt
    + beta_cost * cost_transit
    + beta_time_pt * dur_pt
)

v_driving_Model1 = (
    beta_cost * cost_driving
    + beta_time_driving * dur_driving
)

In [14]:
V_Model1 = {1: v_walk_Model1, 2: v_cycling_Model1, 3: v_PT_Model1, 4: v_driving_Model1}
logprob_Model1 = models.loglogit(V_Model1, None, travel_mode)
biogeme_Model1 = bio.BIOGEME(database, logprob_Model1)
biogeme_Model1.modelName = 'logit_Model1'
res_Model1 = biogeme_Model1.estimate()

In [15]:
print(res_Model1.printGeneralStatistics())

Number of estimated parameters:	8
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4372.911
Final log likelihood:	-4372.911
Likelihood ratio test for the init. model:	-0
Rho-square for the init. model:	0
Rho-square-bar for the init. model:	-0.00183
Akaike Information Criterion:	8761.822
Bayesian Information Criterion:	8813.959
Final gradient norm:	2.8918E-03
Nbr of threads:	8



In [16]:
res_Model1.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
beta_cost         -0.136269      0.013923    -9.787478  0.000000e+00
beta_time_cycling -5.436177      0.492925   -11.028412  0.000000e+00
beta_time_driving -5.794675      0.384636   -15.065358  0.000000e+00
beta_time_pt      -3.127377      0.243891   -12.822864  0.000000e+00
beta_time_walking -8.161128      0.386290   -21.126960  0.000000e+00
constant_cycling  -2.560688      0.150700   -16.991966  0.000000e+00
constant_pt       -0.405922      0.069445    -5.845207  5.059401e-09
constant_walking   1.933359      0.129143    14.970653  0.000000e+00

As before, signs are reasonable. Concernign the values of the coefficents, we firstly notice that beta_cost reamined pretty much the same and it seems reasonable becuase people usually care how much money they spend rather than for what they do spend. The wide range of values we can see on beta_time strenghtens our initial assumption, in fact we see how beta_walking is really low, since walking requires phisycal effort, while the two alternatives that just require the individual's focus (driving and cycling) are pretty similar and this seems fine. In the end we see that beta_time_PT is instead the lowest in absolute value, and this is logic for us since when you are in public transport you are able to do other activities or just realx without having to mantain focus. Again, as before, ASC-walking is still bigger than car, as cycling is still the lowest. The only change we notice is the sign of ASC_pt, which is probably due to the value of the beta_time_pt we previously analyzed. Significance of value is confirmed by the t-test values that are all above the 1.96 threshold

In [17]:
res_Model1.likelihood_ratio_test(res_Model0, 0.01)

LRTuple(message='H0 can be rejected at level 1.0%', statistic=607.9289925585272, threshold=11.344866730144373)

To compare the 2 models we used the likelihood ratio test. This is doable since model 0 is a linear restriction of model 1 (assuming all the alternative specific beta_time equal). Our null-hypotesis is that the resricted model (in this case model 0) is the true model. We set the significance level of the test to 0,01 and as we can see from the result the null-hypotesis is rejected, hence our preferred model is model 1

# Model 2:

- Interaction of age with ASCs:

We select 'age' as socio-economic characteristic since we think it has a stornger correlation with the choice of a transport mode rather than sex, which is the other socio-ecomic characteristic we found in the dataset. In fact, if we assume two individuals have the same age, from our experience, their choice (putting all the other parametres equal) would not probably be different. On the other hand, age can impact the choice both from an availability point of view (i.e. young or really old may not have a driving license) and a physical effort one.

In [18]:
age = Variable('age')

In [19]:
beta_cost = Beta('beta_cost', 0, None, None, 0)
beta_time_walking = Beta('beta_time_walking', 0, None, None, 0)
beta_time_cycling = Beta('beta_time_cycling', 0, None, None, 0)
beta_time_pt = Beta('beta_time_pt', 0, None, None, 0)
beta_time_driving = Beta('beta_time_driving', 0, None, None, 0)
ASC_w = Beta('ASC_w', 0, None, None, 0)
ASC_c = Beta('ASC_c', 0, None, None, 0)
ASC_pt = Beta('ASC_pt', 0, None, None, 0)


In [20]:
interacted_ASC_w = ASC_w * age/99
interacted_ASC_c = ASC_c * age/99
interacted_ASC_pt = ASC_pt * age/99


In [21]:
v_w_Model2_ASC = (
    interacted_ASC_w
    + beta_time_walking * dur_walking
)

v_c_Model2_ASC = (
    interacted_ASC_c
    + beta_time_cycling * dur_cycling
)

v_pt_Model2_ASC = (
    interacted_ASC_pt
    + beta_cost * cost_transit
    + beta_time_pt * dur_pt
)

v_driving_Model2_ASC = (
    beta_cost * cost_driving
    + beta_time_driving * dur_driving
)

In [22]:
V_Model2_ASC = {1: v_w_Model2_ASC , 2: v_c_Model2_ASC, 3: v_pt_Model2_ASC, 4: v_driving_Model2_ASC}
logprob_Model2_ASC = models.loglogit(V_Model2_ASC, None, travel_mode)
biogeme_Model2_ASC = bio.BIOGEME(database, logprob_Model2_ASC)
biogeme_Model2_ASC.modelName = 'logit_Model2_ASC'
res_Model2_ASC = biogeme_Model2_ASC.estimate()

In [23]:
print(res_Model2_ASC.printGeneralStatistics())

Number of estimated parameters:	8
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4615.576
Final log likelihood:	-4615.576
Likelihood ratio test for the init. model:	-0
Rho-square for the init. model:	0
Rho-square-bar for the init. model:	-0.00173
Akaike Information Criterion:	9247.151
Bayesian Information Criterion:	9299.289
Final gradient norm:	4.2811E-03
Nbr of threads:	8



In [24]:
res_Model2_ASC.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
ASC_c             -4.876934      0.374371   -13.027024  0.000000e+00
ASC_pt            -0.998334      0.133146    -7.498069  6.483702e-14
ASC_w              1.531976      0.158595     9.659663  0.000000e+00
beta_cost         -0.150734      0.014566   -10.348652  0.000000e+00
beta_time_cycling -8.105935      0.665889   -12.173110  0.000000e+00
beta_time_driving -5.751921      0.391627   -14.687258  0.000000e+00
beta_time_pt      -3.123986      0.244501   -12.776987  0.000000e+00
beta_time_walking -5.265470      0.197674   -26.637139  0.000000e+00

The estimated signs leaves us satisfied. Moreover, from this estimation we can notice that walking is still the highest among the ASCs, but it is also a lot penalized by its beta_time, which we could expect. Anyway the fact that beta_time_cycling is so low seems pretty strange to us, since we thinik that time has a bigger impact in other contexts such as walking. We can also see that beta_cost has a less signifcant impact on the choices.

- Interaction of age with cost:

When we have to analyze the interaction of the age with one of the variables, we decided to link it with cost, since we think that indviduals may have a differtent perception of money depending on their age. For example, when you're a student you give a different value to money than the one you give them when you've been working for several years.

In [25]:
constant_walking = Beta('constant_walking', 0, None, None, 0)
constant_cycling = Beta('constant_cycling', 0, None, None, 0)
constant_pt = Beta('constant_pt', 0, None, None, 0)
beta_time_walking = Beta('beta_time_walking', 0, None, None, 0)
beta_time_cycling = Beta('beta_time_cycling', 0, None, None, 0)
beta_time_pt = Beta('beta_time_pt', 0, None, None, 0)
beta_time_driving = Beta('beta_time_driving', 0, None, None, 0)
beta_cost_int = Beta('beta_cost_int', 0, None, None, 0)

In [26]:
beta_cost_interacted = beta_cost_int * age/99

In [27]:
v_w_Model2_cost = (
    constant_walking
    + beta_time_walking * dur_walking
)

v_c_Model2_cost = (
    constant_cycling
    + beta_time_cycling * dur_cycling
)

v_pt_Model2_cost= (
    constant_pt
    + beta_cost_interacted * cost_transit    
    + beta_time_pt * dur_pt
)

v_driving_Model2_cost = (
    beta_cost_interacted * cost_driving
    + beta_time_driving * dur_driving
)

In [28]:
V_Model2_cost = {1: v_w_Model2_cost, 2: v_c_Model2_cost, 3: v_pt_Model2_cost, 4: v_driving_Model2_cost}
logprob_Model2_cost = models.loglogit(V_Model2_cost, None, travel_mode)
biogeme_Model2_cost = bio.BIOGEME(database, logprob_Model2_cost)
biogeme_Model2_cost.modelName = 'logit_Model2_cost'
res_Model2_cost = biogeme_Model2_cost.estimate()

In [29]:
print(res_Model2_cost.printGeneralStatistics())

Number of estimated parameters:	8
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4387.311
Final log likelihood:	-4387.311
Likelihood ratio test for the init. model:	-0
Rho-square for the init. model:	0
Rho-square-bar for the init. model:	-0.00182
Akaike Information Criterion:	8790.622
Bayesian Information Criterion:	8842.76
Final gradient norm:	1.8462E-03
Nbr of threads:	8



In [30]:
res_Model2_cost.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
beta_cost_int     -0.287265      0.039548    -7.263774  3.763656e-13
beta_time_cycling -5.430248      0.494069   -10.990875  0.000000e+00
beta_time_driving -5.958971      0.382057   -15.597075  0.000000e+00
beta_time_pt      -3.218752      0.242768   -13.258575  0.000000e+00
beta_time_walking -8.180582      0.386817   -21.148430  0.000000e+00
constant_cycling  -2.566973      0.150922   -17.008640  0.000000e+00
constant_pt       -0.407455      0.069331    -5.876925  4.179564e-09
constant_walking   1.939261      0.129410    14.985393  0.000000e+00

The signs we get are still ok for us, then we can also see that this time beta_cost is lower than before, and this is due to the fact that we are considering the interaction with age. We can also see that the ASCs are still pretty much the same. 

Now it's the tome of testing the models. Since the previous preferred model is Model1, and none of our last two models can be seen as a restriction of Model 1, we cannot apply loglikelihood ratio test directly. We instead want to perfrom a Cox test, where we define a third model, as the sum of model 1 and the interaction term. We then perform the likelihood ratio test between the model 1 and the cox model and then between the interacted model and the cox model. The null hypotesis for each of this test is that the restricted model is the better one. We set the significance level of the test to 0,05. If the cox_model results to be the best in both case we have to look for better models, if the models are not rejected in both case we need to change our test and in the other cases we'll be able to state the preferred model. Of course this has to be done for both the interactions we defined. 

In [31]:
v_w_cox1 = v_walk_Model1 + interacted_ASC_w
v_c_cox1 = v_cycling_Model1 + interacted_ASC_c
v_pt_cox1 = v_PT_Model1 + interacted_ASC_pt
v_driving_cox1 = v_driving_Model1

In [32]:
V_cox1 = {1: v_w_cox1, 2: v_c_cox1, 3: v_pt_cox1, 4: v_driving_cox1}
logprob_cox1 = models.loglogit(V_cox1, None, travel_mode)
biogeme_cox1 = bio.BIOGEME(database, logprob_cox1)
biogeme_cox1.modelName = 'logit_cox1'
res_cox1 = biogeme_cox1.estimate()

In [33]:
print(res_cox1.printGeneralStatistics())

Number of estimated parameters:	11
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4348.347
Final log likelihood:	-4348.347
Likelihood ratio test for the init. model:	-0
Rho-square for the init. model:	0
Rho-square-bar for the init. model:	-0.00253
Akaike Information Criterion:	8718.694
Bayesian Information Criterion:	8790.383
Final gradient norm:	2.3308E-03
Nbr of threads:	8



In [34]:
res_cox1.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
ASC_c             -1.056331      0.365441    -2.890563  3.845519e-03
ASC_pt            -0.959524      0.187752    -5.110597  3.211423e-07
ASC_w             -1.397502      0.236492    -5.909296  3.435729e-09
beta_cost         -0.137210      0.013975    -9.818290  0.000000e+00
beta_time_cycling -5.425166      0.493299   -10.997715  0.000000e+00
beta_time_driving -5.742581      0.383501   -14.974113  0.000000e+00
beta_time_pt      -3.113525      0.243493   -12.786929  0.000000e+00
beta_time_walking -8.186674      0.389671   -21.009213  0.000000e+00
constant_cycling  -2.131890      0.210885   -10.109264  0.000000e+00
constant_pt       -0.012516      0.102428    -0.122195  9.027448e-01
constant_walking   2.502574      0.169588    14.756785  0.000000e+00

In [35]:
print(res_cox1.likelihood_ratio_test(res_Model1, 0.05))
print(res_cox1.likelihood_ratio_test(res_Model2_ASC, 0.05))

LRTuple(message='H0 can be rejected at level 5.0%', statistic=49.127263927997774, threshold=7.814727903251179)
LRTuple(message='H0 can be rejected at level 5.0%', statistic=534.4566876591743, threshold=7.814727903251179)


In this case both model are rejected and so, as we said, we need to keep looking for other models

In [36]:
v_w_cox2 = v_walk_Model1  
v_c_cox2 = v_cycling_Model1
v_pt_cox2 = v_PT_Model1 + beta_cost_interacted * cost_transit
v_driving_cox2 = v_driving_Model1 + beta_cost_interacted * cost_driving

In [37]:
V_cox2 = {1: v_w_cox2, 2: v_c_cox2, 3: v_pt_cox2, 4: v_driving_cox2}
logprob_cox2 = models.loglogit(V_cox2, None, travel_mode)
biogeme_cox2 = bio.BIOGEME(database, logprob_cox2)
biogeme_cox2.modelName = 'logit_cox2'
res_cox2 = biogeme_cox2.estimate()

In [38]:
print(res_cox2.printGeneralStatistics())

Number of estimated parameters:	9
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4371.267
Final log likelihood:	-4371.267
Likelihood ratio test for the init. model:	-0
Rho-square for the init. model:	0
Rho-square-bar for the init. model:	-0.00206
Akaike Information Criterion:	8760.533
Bayesian Information Criterion:	8819.188
Final gradient norm:	8.6372E-04
Nbr of threads:	8



In [39]:
res_cox2.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
beta_cost         -0.194854      0.033653    -5.790010  7.038225e-09
beta_cost_int      0.147775      0.079102     1.868163  6.173939e-02
beta_time_cycling -5.450925      0.493378   -11.048165  0.000000e+00
beta_time_driving -5.797144      0.384984   -15.058141  0.000000e+00
beta_time_pt      -3.124660      0.244302   -12.790159  0.000000e+00
beta_time_walking -8.165163      0.386227   -21.140848  0.000000e+00
constant_cycling  -2.559675      0.150780   -16.976168  0.000000e+00
constant_pt       -0.406721      0.069511    -5.851144  4.882045e-09
constant_walking   1.933683      0.129094    14.978860  0.000000e+00

In [40]:
print(res_cox2.likelihood_ratio_test(res_Model1, 0.05))
print(res_cox2.likelihood_ratio_test(res_Model2_cost, 0.05))

LRTuple(message='H0 cannot be rejected at level 5.0%', statistic=3.288482000320073, threshold=3.841458820694124)
LRTuple(message='H0 can be rejected at level 5.0%', statistic=32.08931386668337, threshold=3.841458820694124)


Preferred model is Model 1

# Model 3: #

We would like to apply the boxcox transofrmation to the time variables. Our choice is based on the fact that we think it's more likely to change the perception of time in a non linear way, when the duration of a travel increases, especially in context such as walking or cycling. Anyway, since we want to be sure that our assumption does not bias the model too much, we are going to apply the boxcox transformation which depending on lambda parameter will give us a wider range of correlation between the time and its perception. 

In [41]:
lambda_boxcox_w = Beta('lambda_boxcox_w', 1, -10, 10, 0)
lambda_boxcox_c = Beta('lambda_boxcox_c', 1, -10, 10, 0)
lambda_boxcox_pt = Beta('lambda_boxcox_pt', 1, -10, 10, 0)
lambda_boxcox_driving = Beta('lambda_boxcox_driving', 1, -10, 10, 0)
boxcox_time_w = models.boxcox(dur_walking, lambda_boxcox_w)
boxcox_time_c = models.boxcox(dur_cycling, lambda_boxcox_c)
boxcox_time_pt = models.boxcox(dur_pt, lambda_boxcox_pt)
boxcox_time_driving = models.boxcox(dur_driving, lambda_boxcox_driving)

In [42]:
v_w_Model3 = (
    constant_walking
    + beta_time_walking * boxcox_time_w
)

v_c_Model3 = (
    constant_cycling
    + beta_time_cycling * boxcox_time_c
)

v_pt_Model3= (
    constant_pt
    + beta_cost * cost_transit    
    + beta_time_pt * boxcox_time_pt
)

v_driving_Model3 = (
    beta_cost * cost_driving
    + beta_time_driving * boxcox_time_driving
)

In [43]:
v_Model3 = {1: v_w_Model3, 2: v_c_Model3, 3: v_pt_Model3, 4: v_driving_Model3}
logprob_Model3 = models.loglogit(v_Model3, None, travel_mode)
biogeme_Model3 = bio.BIOGEME(database, logprob_Model3)
biogeme_Model3.modelName = 'logit_Model3'
res_Model3 = biogeme_Model3.estimate()


In [69]:
print(res_Model3.printGeneralStatistics())

Number of estimated parameters:	12
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4277.532
Final log likelihood:	-4277.532
Likelihood ratio test for the init. model:	-0
Rho-square for the init. model:	0
Rho-square-bar for the init. model:	-0.00281
Akaike Information Criterion:	8579.064
Bayesian Information Criterion:	8657.271
Final gradient norm:	2.7665E-02
Nbr of threads:	8



In [44]:
res_Model3.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
beta_cost             -0.121931      0.013848    -8.805163  0.000000e+00
beta_time_cycling     -3.219416      0.508661    -6.329198  2.464391e-10
beta_time_driving     -3.376682      0.296580   -11.385409  0.000000e+00
beta_time_pt          -3.551234      0.337464   -10.523293  0.000000e+00
beta_time_walking     -5.256316      0.379644   -13.845368  0.000000e+00
constant_cycling      -2.329929      0.260703    -8.937104  0.000000e+00
constant_pt            1.539500      0.140732    10.939190  0.000000e+00
constant_walking       0.099235      0.282442     0.351345  7.253293e-01
lambda_boxcox_c        0.279432      0.129578     2.156483  3.104598e-02
lambda_boxcox_driving  0.320108      0.057551     5.562126  2.665073e-08
lambda_boxcox_pt       0.753032      0.133095     5.657848  1.532827e-08
lambda_boxcox_w        0.306419      0.069256     4.424418  9.670259e-06

The estiamted lambdas are quite different, and this leaves us happy, especially the lambda_boxcox_pt is close to 1 that means it's close to be linear (whihc we expected since on public transport you may not feel too much difference between the first minutes and the last ones, since as we've already said you're able to rest or to do other stuffs). Another interesting fact we can notice is that all lambdas are smaller than 1 which means that the more we increase the time the less impact we get on the individual. The signs of the betas are still coherent and the order of ASCs reamins as usual. Looking at t-test values, we see that all of them are gratetr in absoulte value than 1.96, except for ASC_walking, which results to be quite low meaning that this is not really significant.

In [45]:
res_Model3.likelihood_ratio_test(res_Model1, 0.05)

LRTuple(message='H0 can be rejected at level 5.0%', statistic=190.7572699829543, threshold=9.487729036781154)

We applied the likelihood ratio test since boxcox transformation can be resricted to model 1 (linear) by putting all the lambdas to 1. The result seems good and so now our preferred model in Model3, this confirms our intial assumptions about different perception of time.

# Model 4: #

We choose to test a nested structure. Divding our options in MOTORIZED and UNMOTORIZED (fare grafi su latex). The division seems reasonable to us since when you have to decide a travel mode, at least in our case, the flow of thoughts usually starts from deciding if we want to go with a motorized or unmotorized alternative, depending on how fast do we have to be, on how do we phisically fell and on other factors such as weather conditions. We also notice that this division can be interpreted as a split between FREE and PAID alternatives, which will give us the possibility to interpret the reuslts in a dual way.

In [60]:
mu_motorized = Beta('mu_motorized', 1, 0, None, 0)
mu_unmotorized = Beta('mu_unmotorized', 1, 0, None, 0)
unmotorized = mu_unmotorized, [1, 2]
motorized = mu_motorized, [3, 4]
nests = unmotorized, motorized
logprob_Model4 = models.lognested(v_Model3, None, nests, travel_mode) 
biogeme_Model4 = bio.BIOGEME(database, logprob_Model4)
biogeme_Model4.modelName = 'nested_logit_Model4'
res_Model4 = biogeme_Model4.estimate()


In [61]:
res_Model4.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
beta_cost             -0.095513      0.020288    -4.707939  2.502342e-06
beta_time_cycling     -4.195616      1.134928    -3.696813  2.183231e-04
beta_time_driving     -2.569629      0.541272    -4.747389  2.060589e-06
beta_time_pt          -2.784873      0.552464    -5.040820  4.635424e-07
beta_time_walking     -6.107417      0.570887   -10.698116  0.000000e+00
constant_cycling      -2.780108      0.390406    -7.121066  1.070921e-12
constant_pt            1.190504      0.239824     4.964064  6.903331e-07
constant_walking      -1.634590      0.593654    -2.753439  5.897275e-03
lambda_boxcox_c        1.556361      0.556275     2.797828  5.144752e-03
lambda_boxcox_driving  0.258577      0.067901     3.808145  1.400134e-04
lambda_boxcox_pt       0.663758      0.137734     4.819120  1.441925e-06
lambda_boxcox_w        0.404151      0.072690     5.559907  2.699189e-08
mu_motorized           1.273700      0.212788     5.985770  2.153688e-09
mu_unmotorized         0.506170      0.066942     7.561324  3.996803e-14

Since the estimation for mu_unmotorized is less than one, this nested logit model is not valid, so for the sake of a good comparison with model 3, we are gonna try some other split, hoping to find a valid one. The next try will be about splitting the altenratives in PUBLIC and PRIVATE

In [56]:
mu_private = Beta('mu_private', 1, 0, None, 0)
private = mu_private, [1,2,4]
public = 1.0,[3]
nests = private, public
logprob_Model4_v2 = models.lognested(v_Model3, None, nests, travel_mode)
biogeme_Model4_v2 = bio.BIOGEME(database, logprob_Model4_v2)
biogeme_Model4_v2.modelName = 'nested_logit_Model4_v2'
res_Model4_v2 = biogeme_Model4_v2.estimate()

In [57]:
res_Model4_v2.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
beta_cost             -0.122300      0.014391    -8.498556  0.000000e+00
beta_time_cycling     -3.215516      0.518201    -6.205149  5.464502e-10
beta_time_driving     -3.371131      0.304441   -11.073174  0.000000e+00
beta_time_pt          -3.545232      0.348346   -10.177320  0.000000e+00
beta_time_walking     -5.285215      0.479692   -11.017934  0.000000e+00
constant_cycling      -2.354015      0.346229    -6.799007  1.053424e-11
constant_pt            1.538232      0.141445    10.875124  0.000000e+00
constant_walking       0.070206      0.413226     0.169897  8.650909e-01
lambda_boxcox_c        0.277023      0.135386     2.046170  4.073969e-02
lambda_boxcox_driving  0.317848      0.063035     5.042450  4.596089e-07
lambda_boxcox_pt       0.745826      0.156915     4.753058  2.003630e-06
lambda_boxcox_w        0.305779      0.069682     4.388188  1.142992e-05
mu_private             0.991459      0.084383    11.749530  0.000000e+00

We notice that we did not put any mu_public since there is just one alternative is this nest. Anyway the estimation of mu_private is still less than 1 so we wanna keep on with our exploration, trying to exploit a cross nested model, where the split is between PRIVATE and MOTORIZED alternatives, where CAR is common to both the nests.

In [75]:
mu_motorized = Beta('mu_motorized', 1, 1, None, 0)
mu_private = Beta('mu_private', 1, 1, None, 0)
alpha_motorized_car = Beta('alpha_motorized_car', 0.5, 0, 1, 0)
alpha_private_car = 1 - alpha_motorized_car
alpha_motorized = {
    1 : 0.0,
    2 : 0.0,
    3 : 1.0,
    4 : alpha_motorized_car
}
alpha_private = {
    1 : 1.0,
    2 : 1.0,
    3 : 0.0,
    4 : alpha_private_car
}

nest_motorized = mu_motorized, alpha_motorized
nest_private = mu_private, alpha_private

nests = nest_motorized, nest_private

logprob_Model4_v3 = models.logcnl_avail(v_Model3, None, nests, travel_mode)
biogeme_Model4_v3 = bio.BIOGEME(database, logprob_Model4_v3)
biogeme_Model4_v3.modelName = 'cnl_avail_Model4_v3'
res_Model4_v3 = biogeme_Model4_v3.estimate()


In [77]:
print(res_Model4_v3.printGeneralStatistics())

Number of estimated parameters:	15
Sample size:	5000
Excluded observations:	0
Init log likelihood:	-4270.811
Final log likelihood:	-4270.487
Likelihood ratio test for the init. model:	0.6490347
Rho-square for the init. model:	7.6e-05
Rho-square-bar for the init. model:	-0.00344
Akaike Information Criterion:	8570.973
Bayesian Information Criterion:	8668.731
Final gradient norm:	3.1588E-02
Nbr of threads:	8



In [76]:
res_Model4_v3.getEstimatedParameters()

Value  Rob. Std err  Rob. t-test  Rob. p-value
alpha_motorized_car    0.476237      0.099541     4.784317  1.715696e-06
beta_cost             -0.090863      0.014994    -6.060091  1.360444e-09
beta_time_cycling     -2.329852      0.501330    -4.647347  3.362312e-06
beta_time_driving     -2.318485      0.352242    -6.582079  4.639156e-11
beta_time_pt          -2.307013      0.366284    -6.298418  3.006997e-10
beta_time_walking     -4.126966      0.483955    -8.527574  0.000000e+00
constant_cycling      -2.484570      0.310403    -8.004348  1.110223e-15
constant_pt            0.964207      0.166287     5.798442  6.693386e-09
constant_walking      -0.769463      0.439351    -1.751361  7.988366e-02
lambda_boxcox_c        0.254179      0.155749     1.631975  1.026848e-01
lambda_boxcox_driving  0.336497      0.072112     4.666309  3.066591e-06
lambda_boxcox_pt       0.756503      0.151171     5.004275  5.607279e-07
lambda_boxcox_w        0.286806      0.074222     3.864137  1.114825e-04
mu_motorized           2.223663      0.388956     5.717006  1.084173e-08
mu_private             1.211620      0.133640     9.066297  0.000000e+00

The estimation now leaves us satisfied, both because the theoretical constraints are respected (mu_motorized and mu_private are both grater than 1) and, aboveall, the values are reasonable. In particular, the fact that alpha is estimated at 0.476 is in line with our expectations since we think that there is no particular dominance between the two nests regarding the belonging of CAR alternative. So now it's time to test our model in order to determine if this version is better than our actual preferred one. We apply a logliekelihood ratio test.

In [78]:
res_Model4_v3.likelihood_ratio_test(res_Model3, 0.05)

LRTuple(message='H0 can be rejected at level 5.0%', statistic=14.090893170741765, threshold=7.814727903251179)

So now the preferred model is Model 4, in its last version, where we propose a division between PRIVATE and MOTORIZED alternatives, with CAR belonging to both nests.

# Market shares: #

In [84]:
condition1 = df['age'] >= 45  
condition2 = df['female'] == 1
condition3 = df['age'] < 45
condition4 = df['female'] == 0  

size_S1 = (condition2 & condition3).sum()
size_S2 = (condition2 & condition1).sum()
size_S3 = (condition4 & condition3).sum()
size_S4 = (condition4 & condition1).sum()
S = size_S1 + size_S2 + size_S3 + size_S4
print("Size of S1: ", size_S1, "\nSize of S2: ", size_S2, "\nSize of S3: ", size_S3, "\nSize of S4: ", size_S4)
print(S)

Size of S1:  1663 
Size of S2:  1004 
Size of S3:  1433 
Size of S4:  900
5000


In [86]:
P1 = 2841376
P2 = 1519948
P3 = 2926408
P4 = 1379198
P = P1 + P2 + P3 + P4

w1 = (P1/P)/(size_S1/S)
w2 = (P2/P)/(size_S2/S)
w3 = (P3/P)/(size_S3/S)
w4 = (P4/P)/(size_S4/S)
print("weight of S1: ", w1, "\nweight of S2: ", w2, "\nweight of S3: ", w3, "\nweight of S4: ", w4)

weight of S1:  0.9856918689022015 
weight of S2:  0.8733729419061216 
weight of S3:  1.1781305028128637 
weight of S4:  0.8840744197900654


In [87]:
filters = {
    'female_44_less' : condition2 & condition3,
    'female_44_more' : condition2 & condition1,
    'male_44_less' : condition4 & condition3,
    'male_44_more' : condition4 & condition1
}

In [88]:
weights = {
    'female_44_less' : w1,
    'female_44_more' : w2,
    'male_44_less' : w3,
    'male_44_more' : w4
}

for k,f in filters.items():
    df.loc[f, 'Weight'] = weights[k]

In [89]:
df

trip_id  household_id  person_n  trip_n  travel_mode  purpose  fueltype  \
0           1             0         0       1            4        3         1   
1          13             1         1       1            4        3         1   
2          19             4         0       1            3        3         6   
3          20             5         1       0            4        3         1   
4          39             9         2       0            4        3         1   
...       ...           ...       ...     ...          ...      ...       ...   
4995    81035         17603         1       1            3        1         1   
4996    81040         17604         2       0            3        1         1   
4997    81045         17605         0       1            4        3         1   
4998    81066         17608         0       3            3        3         6   
4999    81073         17610         1       0            4        3         2   

      faretype  bus_scale  survey_year  ...  dur_pt_rail  dur_pt_bus  \
0            1        1.0            1  ...     0.000000    0.055556   
1            5        0.0            1  ...     0.000000    0.122222   
2            5        0.0            1  ...     0.000000    0.312222   
3            5        0.0            1  ...     0.000000    0.062222   
4            5        0.0            1  ...     0.000000    0.225000   
...        ...        ...          ...  ...          ...         ...   
4995         1        1.0            3  ...     0.216667    0.000000   
4996         2        0.0            3  ...     0.000000    0.353333   
4997         5        0.0            3  ...     0.000000    0.045833   
4998         1        1.0            3  ...     0.000000    0.389444   
4999         1        1.0            3  ...     0.000000    0.400000   

      dur_pt_int  pt_interchanges  dur_driving  cost_transit  \
0       0.000000                0     0.059444           1.5   
1       0.000000                0     0.132222           0.0   
2       0.000000                0     0.221667           0.0   
3       0.000000                0     0.117222           0.0   
4       0.000000                0     0.200833           0.0   
...          ...              ...          ...           ...   
4995    0.000000                0     0.392222           2.4   
4996    0.176667                1     0.288889           0.0   
4997    0.000000                0     0.067778           0.0   
4998    0.000000                0     0.193333           1.5   
4999    0.000000                0     0.259167           1.5   

      cost_driving_fuel  cost_driving_ccharge  driving_traffic_percent  \
0                  0.15                   0.0                 0.112150   
1                  0.50                   0.0                 0.065126   
2                  0.56                   0.0                 0.086466   
3                  0.41                   0.0                 0.097156   
4                  0.48                   0.0                 0.378976   
...                 ...                   ...                      ...   
4995               0.98                   0.0                 0.547450   
4996               0.85                   0.0                 0.175000   
4997               0.17                   0.0                 0.024590   
4998               0.61                   0.0                 0.485632   
4999               0.88                   0.0                 0.233655   

        Weight  
0     0.873373  
1     0.884074  
2     0.873373  
3     0.873373  
4     0.985692  
...        ...  
4995  0.884074  
4996  1.178131  
4997  0.873373  
4998  0.985692  
4999  1.178131  

[5000 rows x 33 columns]

In [90]:
sum(df['Weight'])

4999.999999999712

In [91]:
database_weight = db.Database('lpmc_2', df)

In [92]:
Weight = Variable('Weight')
prob_walking = models.cnl_avail(v_Model3, None, nests, 1)
prob_cycling = models.cnl_avail(v_Model3, None, nests, 2)
prob_pt = models.cnl_avail(v_Model3, None, nests, 3)
prob_driving = models.cnl_avail(v_Model3, None, nests, 4)

simulate = {
    'Weight': Weight,
    'Prob. walking': prob_walking,
    'Prob. cycling': prob_cycling,
    'Prob. public transport': prob_pt,
    'Prob. driving': prob_driving
}

biosim = bio.BIOGEME(database_weight, simulate)
simulated_values = biosim.simulate(res_Model4_v3.getBetaValues())


In [93]:
simulated_values

Weight  Prob. walking  Prob. cycling  Prob. public transport  \
0     0.873373       0.602342       0.027870                0.066062   
1     0.884074       0.014400       0.023150                0.261247   
2     0.873373       0.047593       0.064701                0.308856   
3     0.873373       0.058293       0.032770                0.139065   
4     0.985692       0.211386       0.064820                0.353801   
...        ...            ...            ...                     ...   
4995  0.884074       0.008483       0.025399                0.624868   
4996  1.178131       0.014391       0.038959                0.236167   
4997  0.873373       0.500437       0.029926                0.116683   
4998  0.985692       0.011593       0.020539                0.267154   
4999  1.178131       0.002112       0.013566                0.308989   

      Prob. driving  
0          0.303726  
1          0.701204  
2          0.578850  
3          0.769871  
4          0.369993  
...             ...  
4995       0.341249  
4996       0.710483  
4997       0.352955  
4998       0.700715  
4999       0.675333  

[5000 rows x 5 columns]

In [96]:
simulated_values['Weighted Prob. walking'] = simulated_values['Weight'] * simulated_values['Prob. walking']
simulated_values['Weighted Prob. cycling'] = simulated_values['Weight'] * simulated_values['Prob. cycling']
simulated_values['Weighted Prob. public transport'] = simulated_values['Weight'] * simulated_values['Prob. public transport']
simulated_values['Weighted Prob. driving'] = simulated_values['Weight'] * simulated_values['Prob. driving']

market_share_walking = simulated_values['Weighted Prob. walking'].mean()
market_share_cycling = simulated_values['Weighted Prob. cycling'].mean()
market_share_pt = simulated_values['Weighted Prob. public transport'].mean()
market_share_driving = simulated_values['Weighted Prob. driving'].mean()

print("Market share of walking: ", market_share_walking, "\nMarket share of cycling: ", market_share_cycling, "\nMarket share of public transport: ", market_share_pt, "\nMarket share of driving: ", market_share_driving)
print(market_share_cycling + market_share_driving + market_share_pt + market_share_walking)

Market share of walking:  0.17100899220399005 
Market share of cycling:  0.031905523271867214 
Market share of public transport:  0.36380684001252495 
Market share of driving:  0.43327864451161857
1.0000000000000009


In [99]:
biogeme_Model4_v3.bootstrap_samples = 100
results_bootstrap = biogeme_Model4_v3.estimate(run_bootstrap=True)


 80%|████████  | 80/100 [1:47:22<22:48, 68.40s/it]  src/cythonbiogeme/cpp/biogeme.cc:442: Biogeme exception: Error for data entry 2500 : src/cythonbiogeme/cpp/bioExprElem.cc:75: Biogeme exception: Invalid value for expression <log(MultiSum(*(*(^(0*mu_motorized lit[13],free[13] <2500>(2.68325)),exp(*(-(mu_motorized lit[13],free[13] <2500>(2.68325),1),+(constant_walking lit[8],free[8] <2500>(-0.221638),*(beta_time_walking lit[5],free[5] <2500>(-1.31282),Elem[==(dur_walking lit[35], fixed[20] <2500>(0.548056),0)](0:Elem[*(<(lambda_boxcox_w lit[12],free[12] <2500>(0.455017)<1e-05),>(lambda_boxcox_w lit[12],free[12] <2500>(0.455017),-1e-05))](0:/(-(^(dur_walking lit[35], fixed[20] <2500>(0.548056)*lambda_boxcox_w lit[12],free[12] <2500>(0.455017)),1),lambda_boxcox_w lit[12],free[12] <2500>(0.455017));1:+(+(+(log(dur_walking lit[35], fixed[20] <2500>(0.548056)),*(lambda_boxcox_w lit[12],free[12] <2500>(0.455017),^(log(dur_walking lit[35], fixed[20] <2500>(0.548056))*2))),/(*(^(lambda_boxcox_

: 

: 

In [ ]:
betas = biogeme_Model4_v3.freeBetaNames()
b = results_bootstrap.getBetasForSensitivityAnalysis(betas)
left, right = biosim.confidenceIntervals(b, 0.9)

In [ ]:
left['weighted Prob. walking'] = left['Weight'] * left['Prob. walking']
left['weighted Prob. cycling'] = left['Weight'] * left['Prob. cycling']
left['weighted Prob. public transport'] = left['Weight'] * left['Prob. public transport']
left['weighted Prob. driving'] = left['Weight'] * left['Prob. driving']

right['weighted Prob. walking'] = right['Weight'] * right['Prob. walking']
right['weighted Prob. cycling'] = right['Weight'] * right['Prob. cycling']
right['weighted Prob. public transport'] = right['Weight'] * right['Prob. public transport']
right['weighted Prob. driving'] = right['Weight'] * right['Prob. driving']

left_market_share_walking = left['weighted Prob. walking'].mean()
left_market_share_cycling = left['weighted Prob. cycling'].mean()
left_market_share_pt = left['weighted Prob. public transport'].mean()
left_market_share_driving = left['weighted Prob. driving'].mean()

right_market_share_walking = right['weighted Prob. walking'].mean()
right_market_share_cycling = right['weighted Prob. cycling'].mean()
right_market_share_pt = right['weighted Prob. public transport'].mean()
right_market_share_driving = right['weighted Prob. driving'].mean()



In [ ]:
print('CI for market share of walking: [', left_market_share_walking, right_market_share_walking, ']')
print('CI for market share of cycling: [', left_market_share_cycling, right_market_share_cycling, ']')
print('CI for market share of public transport: [', left_market_share_pt, right_market_share_pt, ']')
print('CI for market share of driving: [', left_market_share_driving, right_market_share_driving, ']')